### PR TITLE
[subsystem-benchmarks] Fix availability-write regression tests

### PR DESCRIPTION
**Pull request by AndreiEres**
Adds availability-write regression tests.
The results for the `availability-distribution` subsystem are volatile, so I had to reduce the precision of the test.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5533502

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5541417

File: polkadot/node/network/availability-distribution/tests/availability-distribution-regression-bench.rs [hunk 1]
```diff
@@ -57,13 +58,13 @@ fn main() -> Result<(), String> {
 	println!("{}", usage);
 
 	messages.extend(usage.check_network_usage(&[
-		("Received from peers", 443.333, 0.05),
-		("Sent to peers", 21818.555, 0.05),
+		("Received from peers", 433.333, 0.05),
+		("Sent to peers", 18480.000, 0.05),
```
Comment (sandreim) [inline polkadot/node/network/availability-distribution/tests/availability-distribution-regression-bench.rs:62]: The amount of data sent/received should always be the same if we don't randomize the PoV sizes, or we use the same seed.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5587483

File: polkadot/node/subsystem-bench/src/lib/environment.rs [hunk 2]
```diff
@@ -118,6 +118,7 @@ fn new_runtime() -> tokio::runtime::Runtime {
 		.thread_name("subsystem-bench")
 		.enable_all()
 		.thread_stack_size(3 * 1024 * 1024)
+		.worker_threads(4)
```
Comment (sandreim) [inline polkadot/node/subsystem-bench/src/lib/environment.rs:121]: We need to ensure the number of worker threads chosen here matches:
- the number of CPU cores reserved by K8s/CI for the pod running the bench
- the number of blocking threads used by subsystems to do heavy work

File: polkadot/node/subsystem-bench/src/lib/availability/mod.rs [hunk 3]
```diff
@@ -149,32 +146,21 @@ fn build_overseer_for_availability_write(
...
+) -> (
+	TestState,
+	TestAuthorities,
+	HashMap<H256, Header>,
+	Vec<BlockInfo>,
+	Vec<Vec<Vec<u8>>>,
+	HashMap<H256, Vec<VersionedValidationProtocol>>,
+	NetworkAvailabilityState,
+	MockRuntimeApi,
+) {
```
Comment (sandreim) [inline polkadot/node/subsystem-bench/src/lib/availability/mod.rs:160]: I would use a struct here to improve readability.

File: .gitlab/pipeline/test.yml [hunk 4]
```diff
@@ -503,6 +503,7 @@ subsystem-regression-tests:
     - .run-immediately
   script:
     - cargo bench --profile=testnet -p polkadot-availability-recovery --bench availability-recovery-regression-bench --features subsystem-benchmarks
+    - cargo test --profile=testnet -p polkadot-availability-distribution --test availability-distribution-regression-bench --features subsystem-benchmarks
```
Comment (sandreim) [inline .gitlab/pipeline/test.yml:506]: We should do `cargo bench` here as well

File: polkadot/node/network/availability-distribution/tests/availability-distribution-regression-bench.rs [hunk 5]
```diff
@@ -24,46 +24,73 @@
...
+		.collect();
+	println!("\rDone!{}", " ".repeat(BENCH_COUNT));
+	let average_usage = BenchmarkUsage::average(&usages);
+	println!("{}", average_usage);
 
-	messages.extend(usage.check_network_usage(&[
-		("Received from peers", 443.333, 0.05),
-		("Sent to peers", 21818.555, 0.05),
+	messages.extend(average_usage.check_network_usage(&[
+		("Received from peers", 433.3, 0.05),
```
Comment (sandreim) [inline polkadot/node/network/availability-distribution/tests/availability-distribution-regression-bench.rs:86]: Do we still need having variance of 5% here ? I remember we discussed that this wasn't actually the case as we'd always get the same numbers

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5599791

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5600210

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5604745

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5605193

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5605936

File: polkadot/node/subsystem-bench/src/lib/availability/test_state.rs [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+	pub candidate_hashes: HashMap<CandidateHash, usize>,
+	// Per candidate index receipts.
+	pub candidate_receipt_templates: Vec<CandidateReceipt>,
+	// Per candidate index `AvailableData`
+	pub available_data: Vec<AvailableData>,
+	// Per candiadte index chunks
+	pub chunks: Vec<Vec<ErasureChunk>>,
+	// Per relay chain block - candidate backed by our backing group
+	pub backed_candidates: Vec<CandidateReceipt>,
+	pub block_infos: Vec<BlockInfo>,
```
Comment (sandreim) [inline polkadot/node/subsystem-bench/src/lib/availability/test_state.rs:60]: These fields are missing comments

File: polkadot/node/network/availability-recovery/benches/availability-recovery-regression-bench.rs [hunk 7]
```diff
@@ -38,27 +41,32 @@ fn main() -> Result<(), String> {
...
+		})
+		.collect();
+	println!("\rDone!{}", " ".repeat(BENCH_COUNT));
+	let average_usage = BenchmarkUsage::average(&usages);
+	println!("{}", average_usage);
 
-	messages.extend(usage.check_network_usage(&[
-		("Received from peers", 307200.000, 0.05),
-		("Sent to peers", 1.667, 0.05),
+	// We don't expect any other values for received and sent
```
Comment (sandreim) [inline polkadot/node/network/availability-recovery/benches/availability-recovery-regression-bench.rs:64]: I would rephrase it as no variance is expected here, also, why not 0.0 ?

**Review by sandreim — APPROVED** [review]
Great work! Let's let them be with  `{"build_allow_failure":true}` for a while and see if there is any more flakiness.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5622613

**Review by alexggh — APPROVED** [review]
Looks good to me,

**Comment by dcompoze** [comment]
After this was merged, the CI now fails on TOML format of `polkadot/node/subsystem-bench/Cargo.toml` and also the `continuous-integration/gitlab-subsystem-regression-tests` job seems to fail.
